nifi-docker/dockermaven/Dockerfile

```diff
@@ -26,23 +26,33 @@ ARG NIFI_BINARY
 
 ENV NIFI_BASE_DIR /opt/nifi
 ENV NIFI_HOME $NIFI_BASE_DIR/nifi-$NIFI_VERSION
-
-# Setup NiFi user
-RUN groupadd -g $GID nifi || groupmod -n nifi `getent group $GID | cut -d: -f1` \
-    && useradd --shell /bin/bash -u $UID -g $GID -m nifi \
-    && mkdir -p $NIFI_HOME/conf/templates \
-    && chown -R nifi:nifi $NIFI_BASE_DIR
+ENV NIFI_PID_DIR=${NIFI_HOME}/run
+ENV NIFI_LOG_DIR=${NIFI_HOME}/logs
 
 ADD $NIFI_BINARY $NIFI_BASE_DIR
-RUN chown -R nifi:nifi $NIFI_HOME
+# Setup NiFi user and create necessary directories
+RUN groupadd -g ${GID} nifi || groupmod -n nifi `getent group ${GID} | cut -d: -f1` \
+    && useradd --shell /bin/bash -u ${UID} -g ${GID} -m nifi \
+    && mkdir -p ${NIFI_HOME}/conf/templates \
+    && mkdir -p $NIFI_BASE_DIR/data \
+    && mkdir -p $NIFI_BASE_DIR/flowfile_repository \
+    && mkdir -p $NIFI_BASE_DIR/content_repository \
+    && mkdir -p $NIFI_BASE_DIR/provenance_repository \
+    && mkdir -p $NIFI_LOG_DIR \
+    && chown -R nifi:nifi ${NIFI_BASE_DIR} \
+    && apt-get update \
+    && apt-get install -y jq xmlstarlet procps
 
 USER nifi
 
-# Web HTTP Port & Remote Site-to-Site Ports
-EXPOSE 8080 8181
+# Clear nifi-env.sh in favour of configuring all environment variables in the Dockerfile
+RUN echo "#!/bin/sh\n" > $NIFI_HOME/bin/nifi-env.sh
+
+# Web HTTP(s) & Socket Site-to-Site Ports
+EXPOSE 8080 8443 10000
 
-WORKDIR $NIFI_HOME
+WORKDIR ${NIFI_HOME}
 
 # Startup NiFi
 ENTRYPOINT ["bin/nifi.sh"]
-CMD ["run"]
+CMD ["run"]
```

Comment:
Creating a container with:
`docker run -p 8080:8080 apache/nifi:1.7.0-SNAPSHOT-dockermaven`
results in NiFi starting successfully, but I'm unable to control-c out of the container.  I'm not a docker expert, but I would expect that hitting control-c would kill the container.  Although, since I didn't run it in interactive mode, this is probably a docker thing, not specific to this Dockerfile.

Starting the container with:
`docker run -d -p 8080:8080 apache/nifi:1.7.0-SNAPSHOT-dockermaven`
and then issuing:
`docker logs <container>`
I see the nifi-bootstrap output, but not the nifi-app.log and nifi-user.log output.  Would it be preferable to have this behavior?

Reply:
Ctrl-C works for me, even if it's somewhat slow while nifi is booting (faster when already booted).
The dockermaven image does not contain the wrapper scripts that the dockerhub one does, this is why tailing and sending all logs to stdout doesn't work. I wasn't sure that was intentional or not but I can add it to make the two environments more similar.

Reply:
I tried to use Ctrl-C after NiFi was successfully up and running to kill the container.  I had to open a new shell and use docker kill to bring it down.  Not a big deal, I still think that's due to docker and having started the container in a non-interactive and non-detached method.

For the logging, I'm not sure if there's a reason why the two docker modules have different wrapper scripts.  You could check with @apiri, but most likely it'd be good to bring them in line.  It doesn't have to be done in this PR, though it'd be nice to get this into the NiFi 1.7.0 release if no one disagrees with it.

Reply:
You're right I totally missed that. The biggest problem is that it is even the case with the dockerhub image, which is much more painful. I think I know what the problem is and working on the fix.